keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -0,0 +1,379 @@
+pods that do not have the same imagePullSecret/auths used to successfully pull
+the images in the first place.
+
+This policy change will have no affect on the `pull always` image pull
+policy or for images that are preloaded.
+
+However, for the `pull never` policy if a first pod successfully pulled an image
+with credential and then a second pod with pull never tried to use the image,
+when the feature gate is on the second pod will receive an error message, where
+before and with the feature gate off the second pod would be able to use the image
+pulled with credentials by the first pod.
+
+This new feature will be enabled with a feature gate in alpha. This feature
+improves the security posture for privacy/security of image contents by forcing
+images pulled with an imagePullSecret/auth of a first pod to be re-authenticated
+for a second pod even if the image is already present through the secure pull of
+the first pod.
+
+The new behavior means that if a first pod results in an image pulled with
+imagePullSecrets a second pod would have to also have rights to the image in
+order to use a present image.
+
+This means that the image pull policy alwaysPull would no longer be required in
+every scenario to ensure image access rights by pods.
+
+## Motivation
+
+There have been customer requests for improving upon kubernetes' ability to
+secure images pulled with auth. on a node. Issue
+[#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
+for a while.
+
+To secure images one currently needs to inject `AllwaysPullImages` into pod
+specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+### Goals
+
+Modify the current pullIfNotPresent policy management enforced by `kubelet` to
+ensure the images pulled with a secret by `kublet` since boot. During the
+EnsureImagesExist step `kubelet` will require authentication of present images
+pulled with auth since boot.
+
+Optimize to only force re-authentication for a pod container image when the
+secret used to pull the container image is not present. IOW if an image is
+pulled with authentication for a first pod, subsequent pods that have the same
+authentication information should not need to re-authenticate.
+
+Images already present at boot or loaded externally to `kubelet` or successfully
```

Comment:
Do we have to worry about cases where an image didn't require auth before but has since been updated to require auth?

Reply:
yes, i proposed a configurable timeout in risks and mitigations

we don't have a way to do expirations in the distribution api "yet" .. but either way we could do a timeout.

I'll update the risk section to mention this will be added to beta.